### ISSUE TITLE
UDP support for Freerdp

### ISSUE THREAD

**Issue by gnshkl1** [FreeRDP/FreeRDP#4978]
Needed UDP feature support in Freerdp same as microsoft RDP8

**Comment by akallabeth** [comment]
Relevant specifications are `[MS-RDPEDP]` and `[MS-RDPEDP2]`

**Comment by ThomasCr** [comment]
Any Update on this? It is essential feature when using RD Gateway with Windows Server >= 2012 over small uplinks.

**Comment by hardening** [comment]
I'd really like to work on this if someone wanna finance it...

**Comment by ThomasCr** [comment]
I am am ready to give 50 € from my personal point.

**Comment by monkeytreehosting** [comment]
What would be needed to get this going? Moving forward, this will be a key requirement...

@hardening

**Comment by monkeytreehosting** [comment]
@akallabeth 

I'm relatively new to this, but would I be correct in saying you meant to reference [MS-RDPEUDP] and [MS-RDPEUDP2]

https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeudp/
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeudp2/

Please do correct me if otherwise.

**Comment by akallabeth** [comment]
@monkeytreehosting yes, that are the specs.
Thing is the transport layer will need major modifications for this.

**Comment by ram-t87** [comment]
Hello. Please tell us what amount is required to finalize support for UDP transport. My clients are interested in making rdp work over UDP and I may be able to collect the necessary amount.

**Comment by hardening** [comment]
@ram-t87 you can contact me in private for this.

**Comment by loligans** [comment]
I also would love this very much!

**Comment by nikiwaibel** [comment]
:+1:

**Comment by redpoisonv3** [comment]
i support this it would help a lot to have

**Comment by amarel123** [comment]
Hello @hardening :) Could you please describe the current progress? I have found your article:
https://www.hardening-consulting.com/en/posts/20210131-udp-support-1.html
which probably means work began last year?

**Comment by hardening** [comment]
@amarel123 you can contact me privately for that, cheers.

**Comment by wqlxx** [comment]
> @amarel123 you can contact me privately for that, cheers.

hi @hardening I had send a email about finance the UDP support in RDP to the [contact@hardening-consulting.com](mailto:contact@hardening-consulting.com). Please check your email.

**Comment by tofer523** [comment]
Does this have any updates?

**Comment by mervinning** [comment]
Does this have a plan?

**Comment by indibeningi** [comment]
7 years...

**Comment by akallabeth** [comment]
@indibeningi waiting for you pull request ;)

**Comment by indibeningi** [comment]
Yeah guys, I’m sorry, you’re absolutely right. I shouldn’t be asking for things when this is just a simple open-source community project and nobody is in a position to demand anything from anyone. I apologize.

**Comment by indibeningi** [comment]
I don't have such low-level language skills and I wish I could help, but I’m afraid my knowledge and capabilities are limited. I’ll leave it to you software warriors to one day finally do justice to RDP on Linux.

**Comment by stalker-vishnu** [comment]
Good afternoon everyone,

I have compiled version 3.24.3 with support for UDP and UDP Proxy.
I am also connecting using the +multitransport option, and I am sure that UDP is enabled on the server side (I verified this by connecting from another Windows machine).

However, under all these conditions, UDP is not working on the Linux machine

**Comment by Fantu** [comment]
Hi, the multitransport option might mistakenly lead people to believe that there is full UDP support; unfortunately, that hasn't been done yet.
It's not a quick and easy task; I think it would require someone to fund it, and I suppose the best approach would be to fund someone who is already familiar with freerdp and has already started working on it:
https://www.hardening-consulting.com/en/posts/20210131-udp-support-1.html
https://www.hardening-consulting.com/en/posts/20230109-udp-support-2.html

**Comment by kazenshi** [comment]
I see a lot of comments about funding this project. If the maintainers of this project would be open to funding this project, could someone in a position of authority propose a funding target as well as a place for the community to fund the project? @akallabeth @Fantu

**Comment by Fantu** [comment]
I suppose is better ask to @hardening that started to work on it if a "big" funding is needed and using fundraising services may be good for him

**Comment by klinki** [comment]
Hi guys, first of all, I'd like to thank authors and all contributors of FreeRDP for such a great work!

And now to the point - I desperately needed some high-performant solution for RDP for my mac, so I decided to try vibe-coding UDP support to FreeRDP. I used Facebook Muse Spark 1.3 model for coding and debugging and GPT-6 Astra for code reviews and debugging as well.

And to my surprise, I managed to get it working in quite a short time. I have it in my repo here: https://github.com/klinki/FreeRDP/tree/feat/add-udp

I needed a bit more tweaks to my use so that branch is a bit messy, it contains lot of debug scripts and test outputs etc... And honestly, this is way out of my expertise, so I don't want to directly contribute it as PR (at least not now). BUT, I know it could be helpful and it might be a good source of inspiration for someone who has the knowledge and wants to implement it. Or maybe if someone says it is valuable for contribution, I would consider cleaning that branch up and removing the messy parts and even opening the PR.

Right now, it seems the UDP part is working quite well and I'm currently working on fixing other issues, especially multimonitor setup with different screens using different DPIs.

I have been using the UDP version for couple of hours of continuous work and it worked quite well. Of course there might be still some issues or some parts implemented incorrectly. Once I fix the few more annoyances I have with multimon and DPI issues, I will start using it almost on daily basis so it will be hardened by my own usage.

**Comment by akallabeth** [comment]
@klinki interesting.
can confirm it compiles, but breaks horribly if no `UDP` can be established (e.g. `wireguard` tunnel)

will fire up a local vm and check direct connections...

**Comment by akallabeth** [comment]
ok, same with direct connection, freezes after initial channel migration.

still, the implementation looks to be quite far (implementing the messages on the wire, did not yet check the message handling to the application)

**Comment by klinki** [comment]
Interesting, something must be different with my environment. I tested it against my real desktop PC and against VM and it worked with both.

What version of Windows are you testing against? Maybe it is using different RDP server? I was testing against Windows 11 in both cases - real PC and VM.

Edit: Here are my environments:

Desktop PC: Windows 11 Pro (x64), 25H2, 26200.9445
VM: Windows 11 Pro (arm), 25H2, 26200.9445 (so the same)


Now, when I tried to get version of VM I experienced the freeze too. I had to restart it and then it worked. So yeah, something is wrong there, I will check that later.

**Comment by akallabeth** [comment]
was testing a windows 10 VM
